pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentMessageExpiryMonitor.java

```diff
@@ -77,8 +77,10 @@ public boolean expireMessages(int messageTTLInSeconds) {
 
             cursor.asyncFindNewestMatching(ManagedCursor.FindPositionConstraint.SearchActiveEntries, entry -> {
                 try {
-                    long entryTimestamp = Commands.getEntryTimestamp(entry.getDataBuffer());
-                    return MessageImpl.isEntryExpired(messageTTLInSeconds, entryTimestamp);
+                    // When the time of the delayed message is greater than the time specified by TTL, we should
+                    // give up checking the TTL of the current delayed message, because the time of the delayed
+                    // message has not yet arrived, we cannot delete these messages.
+                    return MessageImpl.isEntryExpired(messageTTLInSeconds, entry.getDataBuffer());
```

Comment:
It's a binary search process. If we return a result depending on a non monotone increasing field(delayTime), the search result normally will go wild.

Reply:
Yes @Jason918 , if the user uses a delayed message here and sets a TTL policy at the same time, the situation you described may exist. But compared to the situation where the TTL policy and delayed messages conflict with each other and cause message loss, I think this cost is acceptable. Or is there any other better way to avoid message loss in this scenario?

Reply:
I am not sure why we need to take care of this case. 
It's also kind of message lost for non-delay message. 
Maybe we should just make it clear for users that the delay time (`deliverAtTime - publishTime`) must be smaller than message TTL.

Reply:
I don't think so. Why do users need to perceive this matter? TTL and delayed messages are two completely independent things.

Reply:
TTL is not set by your user?

Reply:
I see we are solving this only for fix delay messages. But it seems we have no idea if the user are setting fix delay or just random delay. And for random delay, this won't solve anything (it's binary search, we are not going through all the messages), the TTL and delay time still conflicts.

Reply:
I feel that this is an expected behavior, it's better not to set TTL for topics that use delayed messages

Reply:
> I feel that this is an expected behavior, it's better not to set TTL for topics that use delayed messages

+1